jorldy/process.py

```diff
@@ -53,44 +52,47 @@ def manage_process(
     path_queue.put(log_manager.path)
     config_manager.dump(log_manager.path)
 
-    step, print_stamp, eval_thread = 0, 0, None
+    heap = {"step": 0, "run_step": run_step, "wait_thread": False, "wait_process": True}
+    step = 0
+    gath_thread = Thread(
+        target=gather_thread,
+        args=(result_queue, metric_manager, heap, "append"),
+    )
+    gath_thread.start()
     try:
-        while step < run_step:
-            wait = True
-            while wait or not result_queue.empty():
-                _step, result = result_queue.get()
-                metric_manager.append(result)
-                wait = False
-            print_stamp += _step - step
-            step = _step
-            if print_stamp >= print_period or step >= run_step:
-                if (
-                    eval_thread is None
-                    or not eval_thread.is_alive()
-                    or step >= run_step
-                ):
-                    if eval_thread is not None:
-                        eval_thread.join()
-                    agent.sync_in(**sync_queue.get())
-                    statistics = metric_manager.get_statistics()
-                    eval_thread = Thread(
-                        target=evaluate_thread,
-                        args=(agent, step, statistics, eval_manager, log_manager),
-                    )
-                    eval_thread.start()
-                print_stamp = 0
+        while step < heap["run_step"]:
+            agent.sync_in(**sync_queue.get())
+            while heap["wait_process"]:
+                time.sleep(0.1)
+            heap["wait_thread"] = True
+            if step != heap["step"]:
+                step = heap["step"]
+                statistics = metric_manager.get_statistics()
+            heap["wait_thread"] = False
+            score, frames = eval_manager.evaluate(agent, step)
+            statistics["score"] = score
+            print(f"Step : {int(step)} / {statistics}")
+            log_manager.write(statistics, frames, step)
     except Exception as e:
         traceback.print_exc()
-        if eval_thread is not None:
-            eval_thread.terminate()
     finally:
-        if eval_thread is not None:
-            eval_thread.join()
+        gath_thread.join()
 
 
-# Evaluate
-def evaluate_thread(agent, step, statistics, eval_manager, log_manager):
-    score, frames = eval_manager.evaluate(agent, step)
-    statistics["score"] = score
-    print(f"Step : {step} / {statistics}")
-    log_manager.write(statistics, frames, step)
+# Gather
+def gather_thread(queue, target, heap, mode):
```

Comment:
As I try to use it in both manager and async, the variable names (queue, target) are not intuitive and the code flow is a bit awkward.